In [603]:
import pandas as pd
import plotly.express as plt
import plotly.graph_objects as go

# Read the data from the CSV files
file1 = "Industry_including_construction_value_added_ _of_GDP.csv"
file3 = "Labor_force_total.csv"
file4 = "Unemployment_total_ _of_total_labor_force.csv"

df1 = pd.read_csv(file1)
df2 = pd.read_csv(file3)
df3 = pd.read_csv(file4)

In [604]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 24 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Country      68 non-null     object 
 1   2000         68 non-null     object 
 2   2001         68 non-null     object 
 3   2002         68 non-null     float64
 4   2003         68 non-null     float64
 5   2004         68 non-null     float64
 6   2005         68 non-null     float64
 7   2006         68 non-null     float64
 8   2007         68 non-null     float64
 9   2008         68 non-null     float64
 10  2009         68 non-null     float64
 11  Unnamed: 11  0 non-null      float64
 12  2010         68 non-null     float64
 13  2011         68 non-null     float64
 14  2012         68 non-null     float64
 15  2013         68 non-null     float64
 16  2014         68 non-null     float64
 17  2015         68 non-null     float64
 18  2016         68 non-null     float64
 19  2017      

In [605]:
df1.head()

Country  2000  2001  2002  2003  2004  2005  2006  2007  2008  ...  2012  \
0  Ireland  31.5  34.1  35.1  32.3  30.8  30.0  30.1  28.5  25.9  ...  23.9   
1  Germany  27.7  27.1  26.4  26.2  26.4  26.4  27.0  27.2  26.8  ...  27.3   
2  Belgium  24.9  24.3  23.8  23.2  23.0  22.6  22.1  22.3  21.6  ...  20.1   
3  Denmark  23.7  22.9  22.7  22.2  22.1  22.2  22.8  22.2  22.6  ...  20.4   
4   Greece  18.7  19.3  19.4  20.5  20.3  17.8  20.0  18.0  15.7  ...  14.8   

   2013  2014  2015  2016  2017  2018  2019  2020  2021  
0  23.9  24.3  38.2  36.9  35.9  36.7  35.4  37.5  37.8  
1  26.8  27.0  27.1  27.6  27.5  27.3    27  26.6  26.7  
2  19.9  19.7  19.7  19.3  19.2  18.9  19.3  19.5  19.6  
3  20.1  19.8  20.0  20.7  20.7  20.8  20.7    20  19.3  
4  15.1  14.1  14.5  14.5  13.7  13.6  13.4    15  15.3  

[5 rows x 24 columns]

In [606]:
df2.head()

Country           2000           2001           2002           2003  \
0  Ireland   1,862,430.00   1,898,123.00   1,941,828.00   1,984,446.00   
1  Germany  39,865,801.00  40,029,198.00  40,056,637.00  40,236,914.00   
2  Belgium   4,416,525.00   4,315,763.00   4,365,043.00   4,402,418.00   
3  Denmark   2,853,060.00   2,841,507.00   2,868,928.00   2,868,194.00   
4   Greece   4,860,889.00   4,819,343.00   4,881,094.00   4,935,423.00   

   Unnamed: 5           2004           2005           2006           2007  \
0         NaN   2,033,886.00   2,151,141.00   2,241,382.00   2,336,096.00   
1         NaN  40,039,826.00  41,239,999.00  41,693,116.00  41,861,246.00   
2         NaN   4,485,839.00   4,626,679.00   4,646,803.00   4,728,844.00   
3         NaN   2,895,623.00   2,893,154.00   2,917,717.00   2,924,753.00   
4         NaN   5,019,191.00   5,019,255.00   5,043,207.00   5,031,298.00   

   ...           2012           2013           2014           2015  \
0  ...   2,225,018.00   2,247,873.00   2,258,498.00   2,285,833.00   
1  ...  41,853,628.00  42,212,988.00  42,458,390.00  42,660,629.00   
2  ...   4,910,968.00   4,962,050.00   4,978,926.00   4,989,597.00   
3  ...   2,871,734.00   2,850,392.00   2,865,028.00   2,895,364.00   
4  ...   4,927,758.00   4,867,433.00   4,812,543.00   4,788,983.00   

            2016           2017           2018           2019           2020  \
0   2,330,010.00   2,354,153.00   2,397,838.00   2,446,760.00   2,432,234.00   
1  43,567,225.00  43,819,028.00  43,935,038.00  44,433,744.00  43,501,190.00   
2   4,997,187.00   5,086,330.00   5,144,068.00   5,194,405.00   5,167,188.00   
3   2,939,677.00   2,966,464.00   2,990,219.00   3,032,098.00   3,028,252.00   
4   4,774,947.00   4,753,621.00   4,727,112.00   4,735,273.00   4,643,796.00   

            2021  
0   2,551,561.00  
1  43,386,527.00  
2   5,262,812.00  
3   3,070,703.00  
4   4,605,900.00  

[5 rows x 24 columns]

In [607]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Country     68 non-null     object 
 1   2000        68 non-null     object 
 2   2001        68 non-null     object 
 3   2002        68 non-null     object 
 4   2003        68 non-null     object 
 5   Unnamed: 5  0 non-null      float64
 6   2004        68 non-null     object 
 7   2005        68 non-null     object 
 8   2006        68 non-null     object 
 9   2007        68 non-null     object 
 10  2008        68 non-null     object 
 11  2009        68 non-null     object 
 12  2010        68 non-null     object 
 13  2011        68 non-null     object 
 14  2012        68 non-null     object 
 15  2013        68 non-null     object 
 16  2014        68 non-null     object 
 17  2015        68 non-null     object 
 18  2016        68 non-null     object 
 19  2017        68 non-null     obj

In [608]:
df3.head()

Country  2000  2001  2002 2003  2004  2005  2006 2007 2008  ...  2012  \
0  Ireland   4.3   3.7   4.2  4.5   4.5   4.3   4.4    5  6.8  ...  15.5   
1  Germany   7.9   7.8   8.5  9.8  10.7  11.2  10.3  8.7  7.5  ...   5.4   
2  Belgium   6.6   6.2   6.9  7.7   7.4   8.4   8.3  7.5    7  ...   7.5   
3  Denmark   4.5   4.2   4.3  5.4   5.2   4.8   3.9  3.8  3.7  ...   7.8   
4   Greece  11.3  10.8  10.4  9.8  10.6  10.1   8.9  8.4  7.7  ...  24.7   

   2013  2014 2015  2016  2017  2018 2019  2020  2021  
0  13.7  11.9  9.9   8.4   6.7   5.7    5   5.6   6.2  
1   5.2     5  4.6   4.1   3.8   3.4  3.1   3.9   3.6  
2   8.4   8.5  8.5   7.8   7.1     6  5.4   5.6   6.3  
3   7.4   6.9  6.3     6   5.8   5.1    5   5.6   5.1  
4  27.7  26.7   25  23.5  21.4  19.2   17  15.9  14.7  

[5 rows x 24 columns]

In [609]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 24 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Country      68 non-null     object 
 1   2000         68 non-null     object 
 2   2001         68 non-null     object 
 3   2002         68 non-null     object 
 4   2003         68 non-null     object 
 5   2004         68 non-null     object 
 6   2005         68 non-null     object 
 7   2006         68 non-null     object 
 8   2007         68 non-null     object 
 9   2008         68 non-null     object 
 10  2009         68 non-null     object 
 11  Unnamed: 11  0 non-null      float64
 12  2010         68 non-null     object 
 13  2011         68 non-null     object 
 14  2012         68 non-null     object 
 15  2013         68 non-null     object 
 16  2014         68 non-null     object 
 17  2015         68 non-null     object 
 18  2016         68 non-null     object 
 19  2017      

## Data Cleaning and Manipulation

In this step, we have identified some unnamed and empty columns in our datasets, and we will remove them.

In [610]:
def drop_unnamed_columns(df: pd.DataFrame) -> pd.DataFrame:
    """Drops all columns from a pandas DataFrame that contain 'Unnamed' in their name.
    
    Args:
        df: The pandas DataFrame from which the columns will be dropped.
    
    Returns:
        A new pandas DataFrame with the specified columns dropped.
    """
    return df.loc[:, ~df.columns.str.contains('Unnamed')]

# Drop all columns containing 'Unnamed' in their name from each DataFrame
df1 = drop_unnamed_columns(df1)
df2 = drop_unnamed_columns(df2)
df3 = drop_unnamed_columns(df3)


We have identified that some values in our datasets are "..". Therefore, let's convert these values to NaN (Not a Number) first, so that we can easily convert our values to float or int data types later on.

In [611]:
import numpy as np

def replace_dots_with_nan(df: pd.DataFrame) -> pd.DataFrame:
    """Replaces all '..' values in a pandas DataFrame with a NaN value.
    
    Args:
        df: The pandas DataFrame in which the replacements will be made.
    
    Returns:
        A new pandas DataFrame with the '..' values replaced with NaN.
    """
    return df.replace("..", np.nan)

# Replace '..' with NaN in each DataFrame
df1 = replace_dots_with_nan(df1)
df2 = replace_dots_with_nan(df2)
df3 = replace_dots_with_nan(df3)


We have noticed that some values in our datasets are of the 'object' data type. Let's convert them to float values so that we can process them easily. However, the code provided above is not working for df2. Due to the dirty nature of our data, we will only perform the conversion for df3 at the moment.

In [612]:
#lets change the data type of all columns except the first one
df1.iloc[:, 1:] = df1.iloc[:, 1:].astype(float)
#change the data type of all columns except the first one
df3.iloc[:, 1:] = df3.iloc[:, 1:].astype(float)

First, in order to convert our values in df2 to numeric data, we noticed that there are unnecessary decimal points on the right side of the values. We truncated and removed those decimal points. We also observed that some values contain "N" values, and we converted these values to NaN. Finally, we completely removed commas and decimal points, and converted our values to float.

In [613]:
# Remove trailing zeros after dots
df2.iloc[:, 1:] = df2.iloc[:, 1:].applymap(lambda x: str(x).replace(".", "").rstrip(".") if isinstance(x, str) else x)
# Remove last two numbers
df2.iloc[:, 1:] = df2.iloc[:, 1:].applymap(lambda x: str(x)[:-2] if isinstance(x, str) else x)
#find N values
df2.iloc[:, 1:] = df2.iloc[:, 1:].applymap(lambda x: str(x).replace("N", "NaN") if isinstance(x, str) else x)
# Remove non-numeric characters from the values
df2.iloc[:, 1:] = df2.iloc[:, 1:].replace('[,.]', '', regex=True).astype(float)

In order to visualize our values on a map, we would like to merge our datasets with a file containing country codes that we will provide shortly. Therefore, we are preparing our datasets for this task.

In [614]:
def melt_dataframe(df: pd.DataFrame) -> pd.DataFrame:
    """Melts a DataFrame, keeping the 'Country' column fixed.
    
    Args:
        df: The DataFrame to melt.
        
    Returns:
        A melted DataFrame.
    """
    melted_df = pd.melt(df, id_vars=['Country'], value_vars=df.columns[1:])
    return melted_df

# Apply the function to each DataFrame
melted_df = melt_dataframe(df1)
melted_df2 = melt_dataframe(df2)
melted_df3 = melt_dataframe(df3)


In [615]:
melted_df.head()

Country variable  value
0  Ireland     2000   31.5
1  Germany     2000   27.7
2  Belgium     2000   24.9
3  Denmark     2000   23.7
4   Greece     2000   18.7

In [616]:
#change 1st column name to Year
melted_df = melted_df.rename(columns={'variable': 'Year'})
melted_df2 = melted_df2.rename(columns={'variable': 'Year'})
melted_df3 = melted_df3.rename(columns={'variable': 'Year'})

In [617]:
#change 2st column name to Contribution of the construction sector to the GBD %
melted_df = melted_df.rename(columns={'value': 'Contribution of the construction sector to the GBD %'})
#change the second column name to Labor force, total
melted_df2 = melted_df2.rename(columns={'value': 'Labor force, total'})
#change the second column name to Unemployment, total (% of total labor force)
melted_df3 = melted_df3.rename(columns={'value': 'Unemployment, total (% of total labor force)'})

In [618]:
#Read the country codes file
country_codes = pd.read_csv("country_codes.tsv", sep="\t") # Read the country codes file

In [619]:
country_codes.head()

Country Alpha-2 code Alpha-3 code  Numeric
0     Afghanistan           AF          AFG        4
1         Albania           AL          ALB        8
2         Algeria           DZ          DZA       12
3  American Samoa           AS          ASM       16
4         Andorra           AD          AND       20

The given code uses the pandas.merge function to merge different data frames. However, the pandas.concat function is typically used to concatenate multiple DataFrames and is often faster. However, in this case, since the merging is done based on a specific column ('Country'), merge is a more suitable option.

In [620]:
#merge melted_df's and country_codes on Country column
merged_df2 = pd.merge(melted_df2, country_codes, on='Country', how='left')

We noticed that some country names do not match their codes. To address this, we have identified that the codes for these countries generally consist of the first three letters. This applies to all countries except for the United States and the United Kingdom.

In [621]:
#Fill the missing values in the 'Alpha-3 code' column with the first three letters of each country name
merged_df2["Alpha-3 code"].fillna(merged_df2["Country"].str[:3].str.upper(), inplace=True)
#change united states to USA
merged_df2.loc[merged_df2["Country"] == "United States", "Alpha-3 code"] = "USA"
#change united kingdom to GBR
merged_df2.loc[merged_df2["Country"] == "United Kingdom", "Alpha-3 code"] = "GBR"

In [622]:
merged_df2.head()

Country  Year  Labor force, total Alpha-2 code Alpha-3 code  Numeric
0  Ireland  2000           1862430.0           IE          IRL    372.0
1  Germany  2000          39865801.0           DE          DEU    276.0
2  Belgium  2000           4416525.0           BE          BEL     56.0
3  Denmark  2000           2853060.0           DK          DNK    208.0
4   Greece  2000           4860889.0           GR          GRC    300.0

We are performing these steps in order to use them in the machine learning section, which we will apply based on the outputs obtained from the graphs and statistical information that we will examine later.

In [623]:
#merge these files melted_df, melted_df2 and melted_df3
merged_melted_df = pd.concat([melted_df, melted_df2, melted_df3], axis=1)
"""" We were able to use the pandas.concat function because we didn't specifically specify a column in this case.
"""

'" We were able to use the pandas.concat function because we didn\'t specifically specify a column in this case.\n'

In [624]:
# Merge these files: melted_df, melted_df2, and melted_df3
merged_melted_df = pd.concat([melted_df, melted_df2, melted_df3], axis=1)
# Identify duplicate columns
duplicate_columns = merged_melted_df.columns[merged_melted_df.columns.duplicated()]
# Drop duplicate columns
merged_melted_df = merged_melted_df.loc[:, ~merged_melted_df.columns.duplicated()]
# Drop NaN values from merged_melted_df
merged_melted_df = merged_melted_df.dropna()

In [625]:
merged_melted_df

Country  Year  \
0           Ireland  2000   
1           Germany  2000   
2           Belgium  2000   
3           Denmark  2000   
4            Greece  2000   
...             ...   ...   
1489   South Africa  2021   
1490       Thailand  2021   
1492        Turkiye  2021   
1493        Ukraine  2021   
1495  United States  2021   

      Contribution of the construction sector to the GBD %  \
0                                                  31.5      
1                                                  27.7      
2                                                  24.9      
3                                                  23.7      
4                                                  18.7      
...                                                 ...      
1489                                               24.5      
1490                                               34.8      
1492                                               31.1      
1493                                               23.5      
1495                                               17.9      

      Labor force, total  Unemployment, total (% of total labor force)  
0              1862430.0                                           4.3  
1             39865801.0                                           7.9  
2              4416525.0                                           6.6  
3              2853060.0                                           4.5  
4              4860889.0                                          11.3  
...                  ...                                           ...  
1489          23621462.0                                          28.8  
1490          40332578.0                                           1.2  
1492          33317079.0                                          12.0  
1493          20285701.0                                           9.8  
1495         166120890.0                                           5.4  

[1414 rows x 5 columns]

## Visualization and Statistical Analysis.

 We have data on the "Contribution of the construction sector to the Gross Domestic Product (GDP) %," "Total labor force," and "Unemployment, total (% of total labor force)." In order to analyze the "Contribution of the construction sector to the GDP %" in countries, it is necessary to consider the other values of the countries and examine them logically. Here, we will focus on the total labor force of countries and select two countries that are similar to Ireland. We will then display them on a graph.

Our topic is countries, and since we have data on over 60 countries, it becomes quite challenging to examine them on a line graph or bar graph. Therefore, we will use a Choropleth map to visualize the labor force data of countries over the years.

We specifically selected the "matter" color_continuous_scale parameter to effectively visualize the differences between countries. Additionally, we set the range_color parameter to filter the data between 1 million and 10 million to easily demonstrate the proximity of other countries' total labor force to Ireland. Currently, it is not essential for us to compare Ireland with a country that has a labor force of more than 10 million.

In [626]:
fig = plt.choropleth(merged_df2,
                     locations="Alpha-3 code",
                     color="Labor force, total",
                     hover_name="Country",
                     animation_frame="Year",
                     color_continuous_scale="matter", #we expained why we choosed this color_continuous_scale's in the upper cell
                     range_color=[1000000, 10000000]) #we expained why we choosed this range_color's range in the upper cell
fig.update_layout(
    title_text="Labor Force of Countries",
    geo=dict(projection={"type": "natural earth"})
)  
fig.show()

Now, we clearly observed two countries, Finland and Croatia, which closely resemble the labor force changes in Ireland over the years. To demonstrate this effectively, we will proceed with this step.

In [627]:
def find_closest_countries(df, country, year, n=2):
    """Find the countries with the most similar labor force to the given country in a specific year.
    
    Args:
        df: The DataFrame containing the labor force data.
        country: The country to compare with.
        year: The year to consider.
        n: The number of closest countries to return. Default is 2.
        
    Returns:
        A list of the countries with the most similar labor force in the given year.
    """
    df_year = df[df['Year'] == year].copy()  # Filter the data for the given year and make a copy
    target_value = df_year[df_year['Country'] == country]['Labor force, total'].values[0]
    df_year.loc[:, 'difference'] = abs(df_year['Labor force, total'] - target_value)  # Use .loc to avoid SettingWithCopyWarning
    closest_countries = df_year.nsmallest(8+1, 'difference')['Country'].tolist()
    closest_countries.remove(country)
    return closest_countries

# Choose the most recent year in the data
latest_year = merged_df2['Year'].min()

# Find the two countries with the most similar labor force to Ireland in the latest year
closest_countries = find_closest_countries(merged_df2, 'Ireland', latest_year)
print(closest_countries)


['Croatia', 'New Zealand', 'Lithuania', 'Costa Rica', 'Singapore', 'Norway', 'Albania', 'Finland']


As we can see, Finland and Croatia are the two countries that resemble the labor force changes in Ireland. Now, let's focus on these two countries by displaying them on a map of Europe, highlighting only Finland and Croatia.

In [628]:
# Filter the data to include only the countries of interest
countries_of_interest = ["Ireland", "Croatia", "Finland"]
filtered_df = merged_df2[merged_df2["Country"].isin(countries_of_interest)]

# Generate the choropleth map
fig = plt.choropleth(filtered_df,
                     locations="Alpha-3 code",
                     color="Labor force, total",
                     hover_name="Country",
                     animation_frame="Year",
                     color_continuous_scale="matter",  # "haline", "agsunset", "icefire", "matter", "solar", "ice", "curl"
                     range_color=[1000000, 4000000])  # Define the color scale limits
fig.update_layout(
    title_text="Labor Force of Countries",
    geo=dict(
        projection={"type": "natural earth"},
        showlakes=True, 
        lakecolor='rgb(255, 255, 255)', 
        showland=True,
        landcolor='rgb(220, 220, 220)',
        showcountries=True,
        countrycolor='rgb(255, 255, 255)',
        scope="europe"  # Update the scope to 'europe' to focus on Europe
    )
)  
fig.show()


Now, to have a clearer view of the values for the "Contribution of the construction sector to the GDP %," "Total labor force," and "Unemployment, total (% of total labor force)" for these three countries (Ireland, Finland, and Croatia), we will combine their data into the same table. By doing so, we can observe their trends on a line plot.

In [629]:
# Filter the dataframes for the selected countries
df_filtered = melted_df[melted_df['Country'].isin(["Ireland", "Croatia", "Finland"])]
df2_filtered = melted_df2[melted_df2['Country'].isin(["Ireland", "Croatia", "Finland"])]
df3_filtered = melted_df3[melted_df3['Country'].isin(["Ireland", "Croatia", "Finland"])]

# Merge the filtered dataframes
merged_filtered = pd.merge(df_filtered, df2_filtered, on=['Country', 'Year'], suffixes=('_df1', '_df2'))
merged_filtered = pd.merge(merged_filtered, df3_filtered, on=['Country', 'Year'])

# Perform further analysis on the merged dataframe
# ...


We have chosen colors that predominantly represent the flags of the countries to show the trends of the countries.

In [630]:
import plotly.express as px

#unemployment rate comparison 
fig1 = px.line(merged_filtered, x='Year', y='Unemployment, total (% of total labor force)', color='Country',
               title='Unemployment Rate Comparison',
               color_discrete_map={'Ireland': 'green', 'Croatia': 'red', 'Finland': 'blue'})
fig1.show()


#counstruction sector contribution to gdp comparison
fig2 = px.line(merged_filtered, x='Year', y='Contribution of the construction sector to the GBD %', color='Country',
               title='Construction Sector Contribution to GDP Comparison',
               color_discrete_map={'Ireland': 'green', 'Croatia': 'red', 'Finland': 'blue'})
fig2.show()

#total labor force comparison
fig3 = px.line(merged_filtered, x='Year', y='Labor force, total', color='Country',
               title='Total Labor Force Comparison',
               color_discrete_map={'Ireland': 'green', 'Croatia': 'red', 'Finland': 'blue'})
fig3.show()


In [631]:
import plotly.express as px

# Filter the dataframes for the selected countries
countries = ["Ireland", "Croatia", "Finland"]
filtered_data = merged_filtered[merged_filtered['Country'].isin(countries)]

# Generate the heatmap using plotly express
fig = px.imshow(filtered_data.pivot(index='Country', columns='Year', values='Contribution of the construction sector to the GBD %'),
                x=filtered_data['Year'].unique(),
                y=countries,
                color_continuous_scale='matter',
                title='Contribution of the Construction Sector to the GBD % Heatmap')

fig.update_xaxes(title_text='Year')
fig.update_yaxes(title_text='Country')

fig.show()


# Hypothesis Tests

In [632]:
import numpy as np
import scipy.stats as stats

confidence_level = 0.95

# For "Contribution of the construction sector to the GBD %"
data_gbd = df_filtered["Contribution of the construction sector to the GBD %"]
# Calculate the statistical measure
mean_gbd = np.mean(data_gbd)
# Calculate the standard error
std_error_gbd = stats.sem(data_gbd)
# Calculate the confidence interval
confidence_interval_gbd = stats.t.interval(confidence_level, len(data_gbd)-1, loc=mean_gbd, scale=std_error_gbd)

print("Confidence Interval for Contribution of the construction sector to the GBD %:")
print(f"{confidence_interval_gbd[0]} - {confidence_interval_gbd[1]}")

# For "Labor force, total"
data_labor = df2_filtered["Labor force, total"]
# Calculate the statistical measure
mean_labor = np.mean(data_labor)
# Calculate the standard error
std_error_labor = stats.sem(data_labor)
# Calculate the confidence interval
confidence_interval_labor = stats.t.interval(confidence_level, len(data_labor)-1, loc=mean_labor, scale=std_error_labor)

print("Confidence Interval for Labor Force, total:")
print(f"{confidence_interval_labor[0]} - {confidence_interval_labor[1]}")


Confidence Interval for Contribution of the construction sector to the GBD %:
25.20930104101185 - 27.77251714080634
Confidence Interval for Labor Force, total:
2165916.279814146 - 2343784.629276763


These results are used to estimate, with a confidence level you determined (in this case, 95%), the range in which the mean of the relevant dataset will fall.

"Contribution of the construction sector to the GBD %": According to your results, at a 95% confidence level, the contribution of the construction sector to the GDP is expected to be between 25.21% and 27.77%.

"Labor Force, total": According to your results, at a 95% confidence level, the total labor force is expected to be between 2,165,916 and 2,343,784.

To examine the confidence intervals, you can look at the width of these intervals. A wide confidence interval indicates less precision in the estimate because the range in which the true mean of the dataset lies is broader. On the other hand, a narrow confidence interval indicates greater precision in the estimate.

The confidence intervals also provide information about the plausibility of a specific value. For example, if you have a source claiming that the contribution of the construction sector to the GDP is 30%, that claim can be met with skepticism because it falls outside your 95% confidence interval.

Lastly, these confidence intervals can be used to make comparisons between different groups or time periods. For instance, calculating confidence intervals for the labor force of two different countries and examining the degree of overlap can provide insights into whether there is a significant difference between the labor forces of the two countries.

# Independent Two-Sample t-test, One-Sample t-test, One-Way ANOVA, Wilcoxon Signed-Rank Tests:

"Hypothesis Testing: Assessing Assumptions with Sample Data

Hypothesis testing is a statistical method used to determine the validity of a specific assumption (hypothesis) based on sample data. These methods are employed to evaluate the validity of claims made about a population using sample data.

Here are some common hypothesis tests and their applications:

Independent Two-Sample t-test: This test is used to evaluate whether there is a significant difference in means between two distinct groups. For instance, this test can be applied to compare the contributions of the construction sector to the GDP in "Ireland" and "Croatia".

One-Sample t-test: This test is used to assess whether the mean of a single group significantly differs from a specific value. For example, it can be employed to test whether the contribution of the construction sector to the GDP in "Ireland" has a mean of 30.0.

One-Way ANOVA: This test is used to evaluate whether there is a significant difference in means among three or more groups. For instance, it can be applied to compare the contributions of the construction sector to the GDP in "Ireland", "Croatia", and "Finland".

Wilcoxon Signed-Rank Test: This test is used to assess whether there is a significant median difference between two paired or dependent groups. For example, it can be employed to compare the contributions of the construction sector to the GDP in "Ireland" and "Croatia" in terms of their medians.

Hypothesis tests are essential components of statistical analysis and are frequently employed in scientific research and data analysis. They allow us to establish relationships between assumptions and the available data, enabling us to make statistically significant conclusions.

In conclusion, hypothesis testing is a fundamental tool in statistical analysis. By formulating accurate hypotheses and selecting appropriate tests, we can evaluate the statistical significance of our results."

In [633]:
import scipy.stats as stats

# Data Filtering
df_filtered = merged_filtered[merged_filtered['Country'].isin(["Ireland", "Croatia", "Finland"])]

ireland_data = df_filtered[df_filtered["Country"] == "Ireland"]["Contribution of the construction sector to the GBD %"]
croatia_data = df_filtered[df_filtered["Country"] == "Croatia"]["Contribution of the construction sector to the GBD %"]
finland_data = df_filtered[df_filtered["Country"] == "Finland"]["Contribution of the construction sector to the GBD %"]

# Independent two-sample t-test
t_statistic, p_value = stats.ttest_ind(ireland_data, croatia_data, equal_var=False)
print("Independent Two-Sample t-Test Results:")
print("t-statistic:", t_statistic)
print("p-value:", p_value)

# One-Sample t-Test
expected_mean = 30.0
t_statistic, p_value = stats.ttest_1samp(ireland_data, expected_mean)
print("\nOne-Sample t-Test Results:")
print("t-statistic:", t_statistic)
print("p-value:", p_value)

# One-Way ANOVA test
f_statistic, p_value = stats.f_oneway(ireland_data, croatia_data, finland_data)
print("\nOne-Way ANOVA Test Results:")
print("F-statistic:", f_statistic)
print("p-value:", p_value)

# Wilcoxon Signed-Rank Test
statistic, p_value = stats.wilcoxon(ireland_data, croatia_data, alternative="two-sided")
print("\nWilcoxon Signed-Rank Test Results:")
print("Test Statistic:", statistic)
print("p-value:", p_value)


Independent Two-Sample t-Test Results:
t-statistic: 7.717771185099637
p-value: 5.2698547119900345e-08

One-Sample t-Test Results:
t-statistic: 0.8679745397869091
p-value: 0.39521924150332277

One-Way ANOVA Test Results:
F-statistic: 33.92276875460229
p-value: 1.0030474225858036e-10

Wilcoxon Signed-Rank Test Results:
Test Statistic: 0.0
p-value: 4.76837158203125e-07


Independent Two-Sample t-Test: The p-value (p-value: 5.27e-08) is much smaller than 0.05, indicating that this result is statistically significant. This suggests a significant difference in the contribution of the construction sector to the GDP as a percentage between Ireland and Croatia. Since the t-statistic is positive, we can conclude that Ireland's contribution is higher than Croatia's.

One-Sample t-Test: The p-value (p-value: 0.395) is greater than 0.05, indicating that this result is not statistically significant. This indicates that the average contribution of Ireland's construction sector to the GDP as a percentage is not significantly different from the expected average (30%).

One-Way ANOVA: The p-value (p-value: 1.00e-10) is very small, indicating that this result is statistically significant. This indicates a significant difference in the contribution of the construction sector to the GDP as a percentage among Ireland, Croatia, and Finland. However, further analysis is required to determine which groups are significantly different from each other.

Wilcoxon Signed-Rank Test: The p-value (p-value: 4.77e-07) is very small, indicating that this result is statistically significant. However, the Wilcoxon test does not provide information about the magnitude of the difference between the two matched groups; it only indicates that the difference is significant. Here, it is stated that there is a significant difference in the contribution of the construction sector to the GDP as a percentage between Ireland and Croatia.

Overall, these results indicate significant differences in the contribution of the construction sector to the GDP as a percentage among Ireland, Croatia, and Finland. However, further research and information are needed to fully understand the implications of these differences and how to address them.

# chi-square test

The chi-square test is commonly used to test independence or association between two categorical variables. In this case, the data focuses on the continuous variable "contribution of the construction sector to the GDP (%)." Therefore, the chi-square test cannot be directly applied in this context.

However, if you categorize your data in a specific way (e.g., countries exceeding or not exceeding a certain threshold), then a chi-square test might be applicable. However, this depends on the purpose of your analysis and whether categorizing your data in this way makes sense.

Now, in order to apply the chi-square test, we will create three categories based on the contribution of countries to the annual revenue of the construction sector. We will divide them into low contribution, moderate contribution, and high contribution. We will then perform a chi-square test between Ireland and Croatia.

In [634]:
# Define the bin edges
bin_edges = [df_filtered["Contribution of the construction sector to the GBD %"].min(), 
             df_filtered["Contribution of the construction sector to the GBD %"].quantile(0.33), 
             df_filtered["Contribution of the construction sector to the GBD %"].quantile(0.66), 
             df_filtered["Contribution of the construction sector to the GBD %"].max()]

# Define the bin labels
bin_labels = ['low cont', 'med cont', 'high cont']

# Assign the data to bins
df_filtered["Construction_Contribution_Level"] = pd.cut(df_filtered["Contribution of the construction sector to the GBD %"], 
                                                        bins=bin_edges, labels=bin_labels, include_lowest=True)



In [635]:
# Filter the data for Ireland and Croatia
df_filtered_2_countries = df_filtered[df_filtered["Country"].isin(["Ireland", "Croatia"])]

# Create a contingency table
contingency_table = pd.crosstab(df_filtered_2_countries['Country'], df_filtered_2_countries['Construction_Contribution_Level'])

# Perform the Chi-square test
chi2, p, dof, expected = stats.chi2_contingency(contingency_table)

# Print the results
print("Chi-square Test Results:")
print("=========================")
print("Chi-square statistic:", chi2)
print("p-value:", p)


Chi-square Test Results:
Chi-square statistic: 30.028708133971293
p-value: 3.015427417980818e-07


The chi-square test results indicate that there is a significant association between the contribution categories (low, moderate, high) and the countries (Ireland and Croatia).

The chi-square statistic value of 30.028708133971293 suggests a strong relationship between the variables. Additionally, the very small p-value (3.015427417980818e-07) indicates that the observed association is unlikely to be due to chance alone.

Based on these results, we can conclude that the contribution categories and the countries are not independent of each other. In other words, the level of contribution to the annual revenue of the construction sector varies significantly between Ireland and Croatia.

It is important to note that the chi-square test does not provide information about the direction or strength of the association. Further analysis or post-hoc tests may be needed to explore the nature of the relationship between the variables and determine which specific contribution categories contribute more significantly to the observed association.

Overall, these findings highlight the importance of considering the contribution categories when examining the relationship between the construction sector's annual revenue and the countries under study (Ireland and Croatia).

## MACHINE LEARNING

We aim to understand and predict the effects of one or more independent variables on continuous dependent variables in our dataset. Our dependent variables are "Labor force, total" and "Unemployment, total (% of total labor force)", and our independent variable is "Contribution of the construction sector to the GBD %".

Regression analysis is a widely used model in the field of machine learning. It is employed to understand the relationship between a dependent variable and one or more independent variables, make predictions, or assess the effects between variables.

Regression analysis is particularly effective in situations where there is a need to predict or understand a continuous dependent variable. For example, it can be used to estimate the impact of income on housing prices or evaluate the effect of an advertising campaign on sales.

Regression analysis is a model based on statistical techniques that mathematically expresses the relationship between variables in a dataset. It calculates the parameters used to analyze the relationships between variables and predict future values.

Choosing regression analysis provides a solid statistical foundation to analyze relationships, make predictions, and evaluate the effects between variables in your dataset.

However, it is important to select the appropriate regression method based on the requirements of your dataset and the objectives of your analysis.

# Regression Analysis

I couldn't achieve the desired performance with linear ridge and lasso regression models, so through my research, I came across an open-source machine learning library called XGBoost and applied it. XGBoost operates based on the boosting algorithm and is particularly known for delivering high performance in regression and classification problems.

In [636]:
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Define the dependent and independent variables
X = merged_melted_df[["Labor force, total", "Unemployment, total (% of total labor force)"]].values
y = merged_melted_df["Contribution of the construction sector to the GBD %"].values

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# Create and train the XGBoost regression model
model = XGBRegressor()
model.fit(X_train, y_train)

# Make predictions on the training data
y_train_pred = model.predict(X_train)

# Make predictions on the test data
y_test_pred = model.predict(X_test)

# Evaluate the performance of the model
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)
r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

# Print the results
print("XGBoost Regression Results:")
print("========================================")
print("Training Set Mean Squared Error (MSE):", mse_train)
print("Test Set Mean Squared Error (MSE):", mse_test)
print("Training Set R^2 Score:", r2_train)
print("Test Set R^2 Score:", r2_test)


XGBoost Regression Results:
Training Set Mean Squared Error (MSE): 1.8077286946231939
Test Set Mean Squared Error (MSE): 26.348597757380304
Training Set R^2 Score: 0.972288003452305
Test Set R^2 Score: 0.6352510573093788


In [637]:
from sklearn.model_selection import train_test_split, GridSearchCV

# Create the XGBoost regression model
model = XGBRegressor()

# Define the parameter grid for Grid Search
param_grid = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5]
}

# Fit the model with the best parameters using GridSearchCV on the training data
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='neg_mean_squared_error')
grid_search.fit(X_train, y_train)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Make predictions using the best model
y_train_pred = grid_search.predict(X_train)
y_test_pred = grid_search.predict(X_test)

# Evaluate the performance of the model
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)
r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

# Print the results
print("XGBoost Regression Results:")
print("========================================")
print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Training Set Mean Squared Error (MSE):", mse_train)
print("Test Set Mean Squared Error (MSE):", mse_test)
print("Training Set R^2 Score:", r2_train)
print("Test Set R^2 Score:", r2_test)


XGBoost Regression Results:
Best Parameters: {'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 100}
Best Score: -37.755885921725366
Training Set Mean Squared Error (MSE): 17.640016846849154
Test Set Mean Squared Error (MSE): 28.427395883088746
Training Set R^2 Score: 0.7295832679897474
Test Set R^2 Score: 0.6064738363960945


In [638]:
from sklearn.model_selection import cross_val_score
from xgboost import XGBRegressor

# Bağımlı değişken ve bağımsız değişkenleri belirleyin
X = merged_melted_df[["Labor force, total", "Unemployment, total (% of total labor force)"]].values
y = merged_melted_df["Contribution of the construction sector to the GBD %"].values

# XGBoost regresyon modelini oluşturun
model = XGBRegressor()

# 5-katlamalı cross-validation ile modelin performansını değerlendirin
scores = cross_val_score(model, X, y, cv=5, scoring='neg_mean_squared_error')

# Performans metriklerinin ortalamasını alın
mean_mse = -scores.mean()

# Sonucu yazdırın
print("Cross-Validation Results:")
print("=========================")
print("Mean Squared Error (MSE):", mean_mse)


Cross-Validation Results:
Mean Squared Error (MSE): 51.44985499141607
